File: app/services/auth_service.py
import json
import random
from datetime import timedelta
from typing import Optional
from sqlalchemy.ext.asyncio import AsyncSession
from redis.asyncio import Redis
from fastapi import HTTPException, status
from app.core import security
from app.repositories.user_repository import user_repo, UserCreate
from app.schemas.auth import AuthResponse, TokenResponse, UserProfile

OTP_EXPIRY_SECONDS = 300  # 5 mins

def _generate_otp() -> str:
    """Generate a random 6-digit OTP."""
    return str(random.randint(100000, 999999))

class AuthService:
    def __init__(self, db: AsyncSession, redis: Redis):
        self.db = db
        self.redis = redis

    async def send_otp(self, phone: str) -> bool:
        """
        Send OTP to the given phone number.
        
        In production: replace the print statement with MSG91 API call.
        The interface stays the same — only the delivery method changes.
        """
        key = f"otp:{phone}"
        otp_code = _generate_otp()
        await self.redis.set(key, otp_code, ex=OTP_EXPIRY_SECONDS)
        
        # DEV MODE: Print OTP to terminal
        # TODO: Replace with MSG91 when paid — msg91.send_otp(phone, otp_code)
        print(f"\n{'='*50}")
        print(f"[DEV OTP] Phone: {phone}  →  OTP: {otp_code}")
        print(f"{'='*50}\n")
        
        return True

    async def verify_otp(self, phone: str, otp: str, full_name: Optional[str] = None, city: Optional[str] = None, email: Optional[str] = None) -> AuthResponse:
        key = f"otp:{phone}"
        stored_otp = await self.redis.get(key)

        # Accept stored OTP or "123456" fallback for seeded admin account
        if stored_otp:
            if stored_otp != otp:
                raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid or expired OTP")
        else:
            # No OTP in Redis — only allow "123456" as emergency dev fallback
            if otp != "123456":
                raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid or expired OTP")
        
        # OTP is valid, remove it
        await self.redis.delete(key)

        # Check if user exists
        user = await user_repo.get_by_phone(self.db, phone=phone)
        
        is_new_user = False
        if not user:
            # If no full_name/city provided, tell frontend it's a new user
            if not full_name or not city:
                return AuthResponse(success=True, is_new_user=True)
            
            # Register user
            user = await user_repo.create(self.db, obj_in=UserCreate(phone=phone, full_name=full_name, city=city, email=email))
            is_new_user = True

        # Generate tokens
        access_token = security.create_access_token(subject=str(user.id))
        refresh_token = security.create_refresh_token(subject=str(user.id))

        # Build UserProfile — load role from DB
        role_name = "customer"
        permissions = []
        
        if not is_new_user and getattr(user, 'roles', None) and len(user.roles) > 0:
            role_name = user.roles[0].role.name
            if user.roles[0].role.permissions:
                permissions = [p.permission.name for p in user.roles[0].role.permissions]

        profile = UserProfile(
            id=user.id,
            phone=user.phone,
            full_name=user.full_name,
            city=user.city,
            role=role_name,
            permissions=permissions,
            created_at=user.created_at
        )

        tokens = TokenResponse(
            access_token=access_token,
            refresh_token=refresh_token,
            expires_in=security.settings.ACCESS_TOKEN_EXPIRE_MINUTES * 60
        )

        return AuthResponse(success=True, is_new_user=is_new_user, tokens=tokens, user=profile)

File: src/lib/auth-service.ts
export type UserRole = "admin" | "staff" | "customer";

export interface User {
  id: string;
  phone: string;
  role: UserRole;
  fullName?: string;
  city?: string;
}

export interface AuthSession {
  user: User;
  token: string;
}

export const MOCK_OTP = "123456";
const API_BASE = "http://localhost:8000/api/v1";

export class AuthService {
  static async sendOTP(phone: string): Promise<boolean> {
    try {
      const response = await fetch(`${API_BASE}/auth/send-otp`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ phone }),
      });
      return response.ok;
    } catch (error) {
      console.error("Failed to send OTP", error);
      return false;
    }
  }

  static async verifyOTP(phone: string, otp: string): Promise<{ success: boolean; user?: User; isNewUser?: boolean; token?: string }> {
    try {
      const response = await fetch(`${API_BASE}/auth/verify-otp`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ phone, otp }),
      });
      
      const data = await response.json();
      
      if (!response.ok) {
        return { success: false };
      }

      if (data.is_new_user) {
        return { success: true, isNewUser: true };
      }

      return { 
        success: true, 
        isNewUser: false, 
        user: {
          id: data.user.id,
          phone: data.user.phone,
          role: data.user.role as UserRole,
          fullName: data.user.full_name,
          city: data.user.city
        },
        token: data.tokens.access_token
      };
    } catch (error) {
      console.error("Failed to verify OTP", error);
      return { success: false };
    }
  }

  static async registerUser(data: { phone: string; otp: string; fullName: string; city: string }): Promise<{ success: boolean; user?: User; token?: string; isNewUser?: boolean }> {
    try {
      const response = await fetch(`${API_BASE}/auth/verify-otp`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ 
          phone: data.phone, 
          otp: data.otp,
          full_name: data.fullName,
          city: data.city
        }),
      });
      
      const resData = await response.json();
      if (!response.ok) {
        return { success: false };
      }

      return { 
        success: true, 
        user: {
          id: resData.user.id,
          phone: resData.user.phone,
          role: resData.user.role as UserRole,
          fullName: resData.user.full_name,
          city: resData.user.city
        },
        token: resData.tokens.access_token
      };
    } catch (error) {
      console.error("Failed to register user", error);
      return { success: false };
    }
  }
}

